### DISCUSSION TITLE
gps/gpsd/signalk setup on lysmarine-bbn-full-bookworm_2024-06-20-r0235

### DISCUSSION THREAD

**Discussion by atchik6056** [bareboat-necessities/lysmarine_gen#439] [General]
Hi Michael
I am wondering why gpsd is set like in below on freshly installed lysmarine-bbn-full-bookworm_2024-06-20-r0235:
1) two gpsd instances running started by lysgpsd@.service and gpsd.service gpsd.socket
2)  lysgpsd@.service listen 29470 from  /dev/ttyLYS_gps_0 -> ttyACM0 though there is no connection to 29470 defined in signalk but connections to 29471 and 29472do exsist as gps1xxx, gps2xxx (total 4 connections)
3)  gpsd.service gpsd.socket is listening to default 2947 but it is not connecting to any serial port as it is commented out from /etc/default/gpsd and there is no services defined in /etc/systems/system
easy to fix but ...
````
user@lysmarine:~ $ ps -ef| grep gps
gpsd        1057       1  0 12:42 ?        00:00:00 /usr/sbin/gpsd -N -p -S 29470 /dev/ttyLYS_gps_0
gpsd        1584       1  0 12:42 ?        00:00:00 /usr/sbin/gpsd -p
user        3653    3592  0 12:49 pts/0    00:00:00 grep --color=auto gps

root@lysmarine:~# lsusb
Bus 002 Device 003: ID 174c:0862 ASMedia Technology Inc. X862
Bus 002 Device 002: ID 2109:0817 VIA Labs, Inc. USB3.0 Hub             
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 005: ID 222a:0001 ILI Technology Corp. Multi-Touch Screen
Bus 001 Device 004: ID 0406:2814 Fujitsu-ICL Computers 2.4G Composite Devic
Bus 001 Device 009: ID 10c4:ea60 Silicon Labs CP210x UART Bridge
Bus 001 Device 008: ID 1a86:7523 QinHeng Electronics CH340 serial converter
Bus 001 Device 007: ID 0bda:1a2b Realtek Semiconductor Corp. RTL8188GU 802.11n WLAN Adapter (Driver CDROM Mode)
Bus 001 Device 006: ID 1546:01a7 U-Blox AG [u-blox 7]
Bus 001 Device 003: ID 2109:2817 VIA Labs, Inc. USB2.0 Hub             
Bus 001 Device 002: ID 2109:3431 VIA Labs, Inc. Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

root@lysmarine:~# ls -al /dev/serial/by-id
total 0
drwxr-xr-x 2 root root 80 Jul 26 04:48 .
drwxr-xr-x 4 root root 80 Jul 26 04:48 ..
lrwxrwxrwx 1 root root 13 Jul 26 04:48 usb-1a86_USB_Serial-if00-port0 -> ../../ttyUSB1
lrwxrwxrwx 1 root root 13 Jul 26 04:48 usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0 -> ../../ttyUSB0

user@lysmarine:~ $ cat /etc/default/gpsd
# Devices gpsd should collect to at boot time.
# They need to be read/writeable, either by user gpsd or the group dialout.
DEVICES=""
#DEVICES="/dev/ttyLYS_gps_0"

# Other options you want to pass to gpsd
GPSD_OPTIONS="-p"

# Automatically hot add/remove USB GPS devices via gpsdctl
USBAUTO="false"

root@lysmarine:~# systemctl status gpsd.service gpsd.socket
● gpsd.service - GPS (Global Positioning System) Daemon
     Loaded: loaded (/lib/systemd/system/gpsd.service; disabled; preset: enabled)
     Active: active (running) since Fri 2024-07-26 04:48:40 UTC; 8h ago
TriggeredBy: ● gpsd.socket
    Process: 1583 ExecStart=/usr/sbin/gpsd $GPSD_OPTIONS $OPTIONS $DEVICES (code=exited, status=0/SUCCESS)
   Main PID: 1584 (gpsd)
      Tasks: 1 (limit: 8737)
        CPU: 34ms
     CGroup: /system.slice/gpsd.service
             └─1584 /usr/sbin/gpsd -p

Jul 26 04:48:40 lysmarine systemd[1]: Starting gpsd.service - GPS (Global Positioning System) Daemon...
Jul 26 04:48:40 lysmarine systemd[1]: Started gpsd.service - GPS (Global Positioning System) Daemon.

● gpsd.socket - GPS (Global Positioning System) Daemon Sockets
     Loaded: loaded (/lib/systemd/system/gpsd.socket; enabled; preset: enabled)
     Active: active (running) since Fri 2024-07-26 04:48:30 UTC; 8h ago
   Triggers: ● gpsd.service
     Listen: /run/gpsd.sock (Stream)
             [::1]:2947 (Stream)
             127.0.0.1:2947 (Stream)
      Tasks: 0 (limit: 8737)
        CPU: 4ms
     CGroup: /system.slice/gpsd.socket

Jul 26 04:48:30 lysmarine systemd[1]: Listening on gpsd.socket - GPS (Global Positioning System) Daemon Sockets.
root@lysmarine:~# 


oot@lysmarine:~# ls -al /dev/*tty*
...
crw-rw-r--+ 1 root dialout 166,  0 Jul 26 04:48 /dev/ttyACM0
lrwxrwxrwx  1 root root          7 Jul 26 04:48 /dev/ttyLYS_ch340_ -> ttyUSB1
lrwxrwxrwx  1 root root          7 Jul 26 04:48 /dev/ttyLYS_cp210x_0001 -> ttyUSB0
lrwxrwxrwx  1 root root          7 Jul 26 04:48 /dev/ttyLYS_gps_0 -> ttyACM0
crw-------  1 root root      5,  3 Jul 26 04:48 /dev/ttyprintk
crw-rw-r--  1 root dialout   4, 64 Jul 26 04:48 /dev/ttyS0
crw-rw-r--+ 1 root dialout 188,  0 Jul 26 04:48 /dev/ttyUSB0
crw-rw-r--+ 1 root dialout 188,  1 Jul 26 04:48 /dev/ttyUSB1
````
Signalk connections:
gpsd – gpsd on 2947
gps1 -  serial on /dev/ttyLYS_gps_1
gps1_gpsd gpsd on 29471
gps2 -  serial on /dev/ttyLYS_gps_2
gps2_gpsd – gpsd on 29472

**Comment by mgrouch** [comment]
It’s intentional to make most USB GPS devices plug and play. If you know your hardware you can simply turn off the connections that are not being used in your set up. 
Even if you do not turn them off the do not have much of negative impact on performance. They are retried once a minute or so and do not affect performance in any significant way. 
Unfortunately software can’t exactly recognize gps all the time properly as it goes by vendor id and product id of the serial interface chip which might be used in some other devices which are not necessarily gps.

**Reply by atchik6056** [reply]
this is not my point: the initial setup of BBN is not configured correctly for gps devices and is not going to work out of the box due to configuration mismatch in udev rules, systemd config and gpsd deamons setup
to make it working with simple USB attached gps mouse like ublox I had to:
1 - edit /etc/default/gpsd pointing to the right device like /dev/ttyLYS_gps_0
2 - run systemctl enable gpsd.serice gpsd.socket ; systemctl start  gpsd.service gpsd.socket
3 - delete /etc/systemd/system/lysgpsd@.service file
4 = optionally disable all irrelevant gpsX connections in signalk
...if yo do not care then it is another story, just trying to help

**Reply by mgrouch** [reply]
You only needed to do step 1. 
And it is how it’s mentioned in the documentation.

**Reply by mgrouch** [reply]
I think I've fixed it in 2024-08-03 version. The path to gpsdctl was wrong in some scripts.

**Comment by free-x** [comment]
could anybody explain me, why do you need/use gpsd?

**Reply by atchik6056** [reply]
quickly. gpsd is server/service, running on the host and listening to serial/usb (typically) channel(s) and then serving received signal to tcpip/ port. any/many gps client(s) which allowed to connect (from local network or host) can connect and read gps data. so this is instead of direct and (typically exclusive) connection to the serial port(s). 
as an option you can configure signalk to read/listen serial port directly as it understands gps protocol so you do not need gpsd, but serial will be reserved by signalk exclusively.

**Reply by free-x** [reply]
I know, what gpsd is. I don't understand why to use?
SignalK does the same

**Reply by atchik6056** [reply]
Then why you asking here , Mr professor?
